만국박람회 [STEP2-2] unchain, bard

unchain123 opened the pull request:
# 만국박람회 step2-2 withCode
@protocorn93 
안녕하세요 콘~!! 코드로 짜본 스탭2입니다. 
두개 다 봐주셔서 감사합니다
잘 부탁드리겠습니다.

## 고민한점
### `objc`
![](https://i.imgur.com/yh8asQR.png)
- `mainViewController`의 extension을 사용하여 @objc 함수를 사용하려고 했지만, 이러한 에러 메세지가 나타났습니다. 
@objc를 사용하기 위해서는 @objc프로토콜을 따르거나 클래스 내에서 사용해야 한다고 한다해서 `mainViewController`로 다시 옮겨 주었습니다.

### `translatesAutoresizingMaskIntoConstraints`
- 첫번째 뷰에서 flag이미지를 추가할 때 `translatesAutoresizingMaskIntoConstraints = false` 실수로 이 메서드를 추가하지 않고 만들어 줬었는데 오토레이아웃에 문제가 없었습니다. 찾아보니 코드로 Constraints를 추가할 때 입력해주는 메서드라고 나왔습니다. flag가 들어간 위치는 stackView안이어서 따로 지정해준 Constraints가 아니어도 제자리를 잡았었던 것이라고 판단했습니다. 추후에 코드를 검토하는 과정에서 이 메서드가 빠진 것을 확인하고 추가 하여 주었는데 일단은 충돌이 생기지 않아 오토레이아웃의 적용에 문제가 없었지만 나중에 어떤 Constaraints를 추가할지도 모르고 직접 설정해 준 값을 사용하고 싶었기 때문에 추가를 해주었습니다.

## 궁금한점
### `isNavigationBarHidden vs navigationBar.isHidden`
<img src="https://i.imgur.com/DwRplj1.gif" width="300" height="550"/>

- 첨부한 이미지 -> `navigationBar.isHidden = true`를 첫번째 뷰의 viewDidLoad에서 설정해 주고 `viewWillDisappear`에서 `isnavigationBarHidden = false`를 설정해줬을 때 나타난 현상

- 처음 `isNavigationBarHidden`과 `navigationBar.isHidden` 을 혼용해서 사용을 해서 위 이미지와 같은 문제가 발생했습니다  
하지만 [navigationBar](https://developer.apple.com/documentation/uikit/uinavigationcontroller/1621877-navigationbar/)공식문서에서 보니 네비게이션바에 대해 설정을 해줄려면 `isNavigationBarHidden`프로퍼티를 사용하거나 `setNavigationBarHidden(_:animated)` 메소드를 사용하라고 나와있었습니다  
- 이에 대해서 이해는 됬지만, self.navigationController?.navigationBar.isHidden 을 사용해서 네비게이션 바를 숨겨주고 표시해줘도 제대로 나타납니다!
접근자체는 되는 것 같은데 또다른 궁금점이 생겼네요..!
첫번째 뷰컨트롤러에서 viewWillAppear 내부에 self.navigationController?.isNavigationBarHidden = true 를 써주고 
viewWillDisappear 에 self.navigationController?.navigationBar.isHidden = false 를 넣어줘도 잘 작동하는데
반대로,
 viewWillAppear 내부에 self.navigationController?.navigationBar.isHidden = true 를 넣어주고,
viewWillDisappear에 self.navigationController?.isNavigationBarHidden = false 를 넣어주면은 또 안되고 있습니다.  
  
- [블로그](https://zeddios.tistory.com/574)에서는 어떤 것을 사용하던지 접근을 똑같이 해주면은 값이 변경된다고 말하는 것같습니다.  
그래서 네비게이션 바의 접근 자체를 같은 방식으로 접근하면은 변경을 해줄 수 있다는 것을 알겠는데 
왜 첫번째 구현은 되고 두번째 구현은 적용이 안되는지는 전혀 모르겠습니다

bar-d committed:
add: 만국박람회에 필요한 이미지들 추가

unchain123 committed:
Merge pull request #1 from bar-d/ImageAssets

add: 만국박람회에 필요한 이미지들 추가

unchain123 committed:
add: EntryList파일 및 모델 폴더 추가

bar-d committed:
feat: Codable을 채택하여 JSON 데이터와 매칭할 모델 타입 구현
- EntryList 구조체 구현

unchain123 committed:
add: dummyJSON폴더 생성 및 items json파일 추가

bar-d committed:
feat: 스네이크 케이스 또는 축약형인 JSON 키 값을 스위프트의 네이밍에 맞게 변환
- CodingKeys 열거형 CodingKey를 채택하여 구현

bar-d committed:
sytle: EntryList구조체 프로퍼티 및 케이스 네이밍 변경

unchain123 committed:
style: 불필요한 주석 제거

unchain123 committed:
add: expositionUniverselle.json 파일 추가

bar-d committed:
add: ExpoInformation 파일 추가

bar-d committed:
feat: ExpoInformation 구조체 구현

unchain123 committed:
style: 주석변경 및 Foundation제거

bar-d committed:
Merge pull request #2 from bar-d/JSONWithCoustomType

Json with coustom type

bar-d committed:
docs: README.md 파일 수정

unchain123 committed:
Merge branch 'step1' into JSONWithCoustomType

bar-d committed:
docs: README.md 파일 수정

bar-d committed:
add: asset 파일 내 JSON 파일 및 누락된 이미지 파일 추가

unchain123 committed:
refator: asset scale에서 1X파일 제거

bar-d committed:
refactor: JSON 더미파일 및 폴더 삭제

unchain123 committed:
Merge pull request #3 from bar-d/feat-1-imageAsset

이미지 Asset 관련 변경사항

bar-d committed:
test: JSON파일에 대한 테스트 코드 추가

unchain123 committed:
Merge pull request #4 from bar-d/feat-2-ParsingTestCode

Feat 2 parsing test code

unchain123 committed:
refator: 변수에서 상수로 변경

bar-d committed:
add: 폴더 생성 및 분리

bar-d committed:
add: JSONParser 파일 생성

bar-d committed:
feat: JsonParser 및 ParseError 에러 구현

unchain123 committed:
Merge pull request #5 from bar-d/feat-3-JSONParser

JsonParser 구현

unchain123 committed:
feat: numberFormatting하는 함수 구현

bar-d committed:
Merge pull request #6 from bar-d/feat-4-Int+extension

feat: numberFormatting하는 함수 구현

bar-d committed:
style: 폴더 위치 이동

unchain123 committed:
feat: 코드로 루트뷰컨트롤러 지정

- 메인스토리보드내 뷰 컨트롤러 제거

bar-d committed:
feat: scrollView 및 stackview 오토레이아웃 설정

unchain123 committed:
feat: titleLabel 구현

bar-d committed:
feat: 첫번째 스토리보드 내 이미지 추가

unchain123 committed:
feat: 첫번째 스토리보드 내 레이블 추가

- 개최지, 방문객수, 개최기간, 만국박람회설명레이블 추가

bar-d committed:
feat: 네비게이션 컨트롤러 및 버튼 스택뷰 구현

unchain123 committed:
refactor: extension 함수분리

bar-d committed:
style: 뷰 컨트롤러 파일 생성 및 네이밍 변경
- 메인 스토리보드 파일 삭제

bar-d committed:
docs: README.md 파일 수정

unchain123 committed:
feat: 버튼에 화면전환 함수 추가

bar-d committed:
feat: 두번째 화면내 내비게이션 구현 및 테이블 뷰 생성
- UITableViewDelegate, UITableViewDataSource 프로토콜 채택

unchain123 committed:
add: CustomTableViewCell파일 생성

bar-d committed:
add: 폴더 생성 및 재정렬
- info.plist 경로 재설정

unchain123 committed:
feat: customTableViewCell구현

bar-d committed:
feat: DescriptionViewController 구현

unchain123 committed:
Merge pull request #10 from bar-d/feat-4-2-controller

Feat 4 2 controller

unchain123 committed:
feat: DataRepository프로토콜 생성

bar-d committed:
refactor: JSONParser 구조체 파싱하는 방법 변경
- JSONParser 구조체 내 DataRepository 프로토콜 채택
- JSONParser 구조체 리팩토링
- NameSpace 열거형 생성해 magic literal, magic number 관리

unchain123 committed:
refactor: JSONPaser 리펙토링

unchain123 committed:
refactor: 메인뷰컨트롤러 extension구현

unchain123 committed:
refactor: 메인뷰컨트롤러내의 함수 extension으로 분리

bar-d committed:
add: EntryListViewController+extension 파일 추가

bar-d committed:
refactor: EntryListViewController 내 함수 리팩토링
- 함수 extension으로 분리
- 데이터 전달

bar-d committed:
refactor: CustomTableViewCell 오토레이아웃 설정

unchain123 committed:
refactor: 세번째 뷰컨트롤러 모델 리펙토링

unchain123 committed:
refactor: 테이블뷰 오토레이이아웃 수정

bar-d committed:
style: 전반적인 컨벤션 수정

unchain123 committed:
Merge pull request #11 from bar-d/refactor-2-2-model

Refactor 2 2 model

unchain123 committed:
refactor: 오토레이아웃 재설정

bar-d committed:
refactor: 네비게이션 바 타이틀 리팩토링

bar-d committed:
refactor: 네비게이션 컨트롤러 바 리팩토링
- 공식문서에 나와있는 프로퍼티 적용

protocorn93 commented on Expo1900/Expo1900/Controller/Cell/CustomTableViewCell.swift:
```diff
@@ -0,0 +1,76 @@
+
+import UIKit
+
+class CustomTableViewCell: UITableViewCell {
```
네이밍이 모호한 것 같네요. 도메인에 맞는 적절한 네이밍으로 변경해주면 좋을 것 같아요.

protocorn93 replied:
이 클래스의 내부 요소들의 접근제어자들도 한번 확인 부탁드릴게요

protocorn93 commented on Expo1900/Expo1900/Controller/Cell/CustomTableViewCell.swift:
```diff
@@ -0,0 +1,76 @@
+import UIKit
+
+class CustomTableViewCell: UITableViewCell {
+    static let identifier = ExpoNameSpace.cellIdentifier.name
```
```swift 
protocol ReusableView: class {}

extension ReusableView where Self: UIView {
    static var reuseIdentifier: String {
        return String(describing: self)
    }
}

extension UITableViewCell: ReusableView { }

CustomTableViewCell.reuseIdentifier
```

많은 분들이 편의성을 위해 위와 같은 방식으로 reuse identifier를 제공하곤합니다! 한번 사용해보시는 것도 도움이 될 것 같아요

unchain123 replied:
이부분은 아직 공부가 안된 부분인것 같아 따로 수정은 하지 않았고 따로 공부해보도록 하겠습니다!

protocorn93 commented on Expo1900/Expo1900/Controller/DescriptionViewController.swift:
```diff
@@ -0,0 +1,28 @@
+   
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.view.backgroundColor = .white
```
이 코드도 `setUI`에 들어갈 수 있을 것 같은데요!

protocorn93 commented on Expo1900/Expo1900/Controller/EntryListViewController.swift:
```diff
@@ -0,0 +1,82 @@
+                   didSelectRowAt indexPath: IndexPath) {
+        let descriptionViewController = DescriptionViewController()
+        
+        descriptionViewController.entryList = entryList?[indexPath.row]
```
이제는 생성자를 통해 넣어주는 것이 더 바람직해보여요!

unchain123 replied:
생성자를 만들어서 처리해 주었습니다.

protocorn93 commented on Expo1900/Expo1900/Controller/EntryListViewController.swift:
```diff
@@ -0,0 +1,82 @@
+    
+    func tableView(_ tableView: UITableView,
+                   cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: ExpoNameSpace.cellIdentifier.name,
```
해당 셀의 identifier는 해당 셀이 관리하는게 맞다고 생각해요.

protocorn93 commented on Expo1900/Expo1900/Controller/EntryListViewController.swift:
```diff
@@ -0,0 +1,82 @@
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: ExpoNameSpace.cellIdentifier.name,
+                                                       for: indexPath) as? CustomTableViewCell else { return UITableViewCell() }
+                
+        tableView.addSubview(cell)
```
이 코드는 불필요한 코드에요! 테이블 뷰는 델리게이트 메서드를 통해 사용자에게 셀 객체만 요청하고 이를 테이블 뷰에 붙이는 것은 테이블 뷰가 내부적으로 진행합니다!

unchain123 replied:
삭제 했습니다.

protocorn93 commented on Expo1900/Expo1900/Controller/EntryListViewController.swift:
```diff
@@ -0,0 +1,82 @@
+                
+        tableView.addSubview(cell)
+        cell.addSubviews()
+        cell.setConstraints()
```
이 코드는 cell의 init 메서드 안에서 호출되어야 하는게 맞아요!

protocorn93 commented on Expo1900/Expo1900/Model/Extension/mainViewController+extension.swift:
```diff
@@ -0,0 +1,147 @@
+        self.setFlagImage()
+    }
+    
+    func setting() {
```
이 메서드의 네이밍만보면 어떤 역할을 하는지 파악하기 어려운데요!

protocorn93 commented on Expo1900/Expo1900/Controller/EntryListViewController.swift:
```diff
@@ -0,0 +1,82 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.view.backgroundColor = .white
+        self.tableView.delegate = self
+        self.tableView.dataSource = self
```
이 두 코드는 `setTableView` 안에 들어갈 수 있을 것 같아요

protocorn93 commented on Expo1900/Expo1900/Model/Extension/EntryListViewController+extension.swift:
```diff
@@ -0,0 +1,46 @@
+import UIKit
+
+extension EntryListViewController {
+    func updateUI() {
+        self.tableView.dataSource = self
+        self.tableView.delegate = self
+        self.navigationController?.isNavigationBarHidden = false
+    }
```
updateUI라는 네이밍이라 무언갈 갱신해주는 메서드라고 생각을 했는데, 그와는 다르네요! 

`setDelegates`, `setNavigation`과 같은 늬앙스의 네이밍으로 분리 해주면 좋을 것 같아요.

unchain123 replied:
분리후 setUI로 변경했습니다.

protocorn93 commented on Expo1900/Expo1900/Controller/MainViewController.swift:
```diff
@@ -0,0 +1,57 @@
+import UIKit
+
+final class MainViewController: UIViewController {
+    let scrollView = UIScrollView()
+    let contentView = UIView()
+    let stackView = UIStackView()
+    let expoTitle = UILabel()
+    let posterImage = UIImageView()
+    let visitorsLabel = UILabel()
+    let locationLabel = UILabel()
+    let durationLabel = UILabel()
+    let descriptionLabel = UILabel()
+    let buttonStackView = UIStackView()
+    let leftFlagImage = UIImageView()
+    let rightFlagImage = UIImageView()
+    let goToKoreaEntryButton = UIButton()
```
다른 뷰 컨트롤러에서 정의한 것 처럼 속성은 선언부에 정의해주는게 더 가독성이 좋을 것 같아요. 저도 속성은 선언부에 정의하고 `set-` 메서드에는 레이아웃이나 데이터만 넣어주고 있어요! 

```swift
let someView: SomeView = {
  let view = SomeView() 
  view.property = value
  return view
}()
```

unchain123 replied:
클로저 형태로 변경했습니다.

protocorn93 commented on Expo1900/Expo1900/Model/Extension/mainViewController+extension.swift:
```diff
@@ -0,0 +1,147 @@
+    }
+    
+    private func setStackViewSubviews() {
+        self.stackView.addArrangedSubview(expoTitle)
+        self.stackView.addArrangedSubview(posterImage)
+        self.stackView.addArrangedSubview(visitorsLabel)
+        self.stackView.addArrangedSubview(locationLabel)
+        self.stackView.addArrangedSubview(durationLabel)
+        self.stackView.addArrangedSubview(descriptionLabel)
+        self.stackView.addArrangedSubview(buttonStackView)
```
스택뷰에 추가될 뷰가 길어지면 아래와 같이 간결하게 작성할 수도 있을 것 같아요. (취향 차이!)

```swift 
[label, imageView, button].forEach { 
  self.stackView.addArrangedSubview($0)
}
```

protocorn93 reviewed (changes requested):
두 분 두 가지 버전으로 코드를 작성해보셨는데 잘 구현해주신 것 같아요! 👏

다만 물리적인 코드 구조에 대해서 이야기해보고 싶은데요. 역할별로 메서드를 나누는 것은 매우 중요해요. 특히 비즈니스로직은 더더욱 그렇구요. 

하지만 UI 코드는 특히 레이아웃 코드는 한 곳에 모여있을 때 읽기 수월해요. 코드 리뷰를 한다고했을 때 컴포넌트 단위로 한 곳에 모여있을 때 읽기 수월하거든요. 이 요소와 저 요소가 어떤 레이아웃 관계를 맺는지 각각 메서드로 점프해서 확인하면 전체적인 레이아웃을 머리속으로 그리기 힘든 부분도 있거든요. 

다만 이건 취향의 문제이고 스타일 차이의 문제이기도해요. 그러니 같이 업무를 진행하는 동료와 충분한 이야기를 나누면서 맞춰가며 작업을 해주시면 좋을 것 같아요! 

## 궁금한 점 

> isHidden vs setNavigationBarHidden/isNavigationBarHidden

`navigationBar`의 `isHidden`은 `UINavigationBar`의 프로퍼티가 아니에요. `UIView`의 프로퍼티이죠. `UINavigationBar`는 `UIView`를 상속받은 클래스이구요!

`UINavigationBar`가 내부적으로 해당 값을 어떻게 오버라이드하고 있는지는 알 수 있는 방법은 없어요! 분명 내비게이션 중에 발생할 수 있는 애니메이션/트랜지션에 맞게 오버라이드 해놓지 않았을까 생각해요. 그렇기에 이를 숨기는데 사용하는 별도 프로퍼티(`isNavigationBarHidden`)와 메서드(`setNavigationBarHidden`)를 열어둔 것이구요. 

그렇기에 사용하라고 열어준 인터페이스를 사용하는 것이 일관적이고 예상치 못한 동작을 방지할 수 있을 것 같아요. (이후에 iOS 버전이 올라가면서 내부적으로 로직이 변경되어도 열어둔 인터페이스를 사용한다면 문제도 없을 것이구요. 문제가 있더라도 deprecated를 붙여줘서 이후 문제가 발생할 수 있다는 것을 직접 알려주기도 해요! 하지만 `isHidden` 프로퍼티를 쓰면 그렇지 못하겠죠!)

bar-d committed:
refactor: CustomCell 네이밍 변경 및 접근 제한자 설정
- CustomTableViewCell에서 EntryListCell로 네이밍 변경
- EntryListCell 내부 함수 접근제한자 설정

unchain123 committed:
refactor: 함수 역할별 메서드 나누기

bar-d committed:
refactor: mainViewController+extension 내부 함수 리팩토링
- updateUI 함수 setUI 로 네이밍 변경
- 스택뷰 내부에 뷰를 추가해주는 함수 forEach 고차함수를 사용하여 추가
- setting() 함수 내부에 있던 설정들 함수 세분화

unchain123 committed:
refactor: 이니셜라이저 적용

bar-d committed:
refactor: 상수 선언부 내 리팩토링 및 함수 분리
- MainViewController내 상수 내부에 클로저를 통한 설정 처리
- esxtension 파일 내부에 함수 분리

unchain123 committed:
style: 함수명 변경

unchain123 commented:
안녕하세요 콘 

리뷰 주신 부분 참고해서 피드백 했습니다. 

>https://github.com/yagom-academy/ios-exposition-universelle/pull/188#discussion_r906746923
이 부분은 아직 공부가 부족한 것 같아 나중에 따로 공부해보려고 합니다.
감사했습니다. 콘
